c2rust-analyze/src/context.rs

```diff
@@ -287,6 +287,8 @@ pub struct GlobalAnalysisCtxt<'tcx> {
     pub addr_of_static: HashMap<DefId, PointerId>,
 
     pub adt_metadata: AdtMetadataTable<'tcx>,
+
+    pub foreign_mentioned_tys: HashSet<Ty<'tcx>>,
```

Comment:
I feel like the current name is fine

Reply:
As @spernsteiner said in https://github.com/immunant/c2rust/pull/960#discussion_r1229915402, this should probably be computed recursively.  What we want to be `FIXED` is any type that ever crosses an FFI boundary, not just types that are used in foreign items (shallowly).  `rustc` generally uses "foreign" to mean "in an `extern` block", but this should include types that are not in `extern` blocks as well, as they still cross an FFI boundary.

Reply:
besides repr C or extern blocks what would constitute crossing an FFI boundary?

Reply:
`#[repr(C)]` doesn't necessarily mean it should be `FIXED`, and all types from `c2rust transpile` will be `#[repr(C)]`.  `#[repr(C)]` still allows more complex, safe types, so we want to rewrite those, just not the ones that actually cross the FFI boundary.  `extern` blocks indicate they will cross the FFI boundary as an import, and as for exports, I think `extern "C"`, too.  The problem is that all `fn`s from `c2rust transpile` will be marked `extern "C"`.  Maybe a better way to detect exports is `#[no_mangle]`.  I'm not sure if mangle `fn`s will ever be called over an FFI boundary or if Rust's name mangling is stable and you are allowed to do that or not.  If something like [the `#[export]` RFC](https://github.com/rust-lang/rfcs/pull/3435) lands, that might be good, but it hasn't even been accepted as an RFC much less implemented and stabilized yet.

Reply:
Given the ambiguity here and the way c2rust currently transpiles, I’m just going to focus on the extern imports and recursing properly and creating issues for the rest.

Reply:
> Maybe a better way to detect exports is `#[no_mangle]` (rather, `#[no_mangle]` and `extern` non-`Rust`).

I think we should ignore the library use case for now and assume that Rust functions are never called from C.  `c2rust-transpile` makes all non-`static` C functions into `#[no_mangle]` Rust functions, so there are a lot of functions in lighttpd that are `#[no_mangle]` just because they're called from a different compilation unit.

> I'm not sure if mangle `fn`s will ever be called over an FFI boundary or if Rust's name mangling is stable and you are allowed to do that or not.

I think we should assume this doesn't happen (regardless of what we decide on the previous point), even though it's technically possible and might even be well-defined.

>besides repr C or extern blocks what would constitute crossing an FFI boundary?

One way to formulate this is in terms of typed expressions/declarations:

* All arguments and return values of `extern`-block `fn`s are FFI-exposed.
* All `extern`-block `static`s are FFI-exposed.
* If `x: *const T` or `x: *mut T` is FFI-exposed, then `*x: T` is also FFI-exposed.
* If `s: S` is FFI-exposed, `S` is a struct type, and `struct S` is declared with `#[repr(C)]`, then each field of `S` is FFI-exposed.
* If `x: T` is FFI-exposed, then all pointers that appear in `T` must be marked `FIXED`.  ("A pointer appears in T" roughly means that if you print `T`, you will see the pointer.  Two pointers appear in `*mut *const i32`; no pointers appear in the type `MyStruct`, regardless of what fields it has.)

Reply:
> I think we should ignore the library use case for now and assume that Rust functions are never called from C. `c2rust-transpile` makes all non-`static` C functions into `#[no_mangle]` Rust functions, so there are a lot of functions in lighttpd that are `#[no_mangle]` just because they're called from a different compilation unit.

I forgot most are `#[no_mangle]`, too.  I agree, then, we should just ignore the export case.  It may not be unreasonable to require users to mark their exports with something like `#[c2rust_analyze_test::fixed_signature]` or `#[c2rust_analyze::export]`.

> * If `x: *const T` or `x: *mut T` is FFI-exposed, then `*x: T` is also FFI-exposed.

I'm not sure what `*x: T` means here.

> * If `s: S` is FFI-exposed, `S` is a struct type, and `struct S` is declared with `#[repr(C)]`, then each field of `S` is FFI-exposed.

Shouldn't this apply to all compound types?  We could do `struct`s only for now if it's simpler, but shouldn't we do it for all compound types, i.e. everything `rustc`'s FFI-safety lint recurses into?

Also, I don't think whether or not the type is `#[repr(C)]` matters.  The field types are still contained in the `struct` and cross the FFI boundary all the same.

Reply:
> Shouldn't this apply to all compound types? We could do `struct`s only for now if it's simpler, but shouldn't we do it for all compound types, i.e. everything `rustc`'s FFI-safety lint recurses into?

Oh, yes, I suppose we need similar rules for arrays and for unions.

Are there any other FFI-safe aggregates?  I know tuples are always implicitly `#[repr(Rust)]`.

> Also, I don't think whether or not the type is `#[repr(C)]` matters. The field types are still contained in the `struct` and cross the FFI boundary all the same.

If the struct isn't either `#[repr(C)]` or #[repr(transparent)]`, then there are no layout guarantees, and the C code can't (or at least really shouldn't) access any fields.

Reply:
> Are there any other FFI-safe aggregates? I know tuples are always implicitly `#[repr(Rust)]`.

From [`ImproperCTypesVisitor::check_type_for_ffi`](https://doc.rust-lang.org/stable/nightly-rustc/src/rustc_lint/types.rs.html#893), it seems that we should recurse into:
* `struct`s
* `union`s
* `enum`s
* references
* raw pointers
* arrays
* `fn` ptrs

Do we already treat non-crate-local types as `FIXED`?

Reply:
> Do we already treat non-crate-local types as `FIXED`?

Roughly, yes.  IIRC we don't even assign `PointerId`s to fields of non-local types, since we can't rewrite them.

Reply:
Okay, so for this recursion, we should make sure to stop eagerly at non-crate-local types.

Reply:
Let me know if 795438ae doesn't accomplish what's desired